Update terraform testing docs and chatops (#578)
cloudposse/docs: content/docs/community/automated-testing.md

## Terraform Testing

> Nearly all of our Terraform modules are updated with automated tests. We have two sets of checks: 

Reviewer: ```suggestion
Most of our Terraform modules are validated with automated tests. We have two sets of checks: 
```

> - First is running terraform module's [`feature-branch workflow`](https://github.com/cloudposse/github-actions-workflows-terraform-module/blob/main/.github/workflows/feature-branch.yml) 
> which does basic sanity checks: linting, formatting and docs. These checks are cheap to run and do not require any special permissions, so we run them automatically on every commit.

Reviewer: ```suggestion
- **[`feature-branch workflow`](https://github.com/cloudposse/github-actions-workflows-terraform-module/blob/main/.github/workflows/feature-branch.yml)**: Runs the terraform module's basic sanity checks: linting, formatting and documentation. These checks are cheap to run and do not require any special permissions, so we run them automatically on every commit.
```

> - Second we have terraform tests based on the [`terratest`](https://github.com/gruntwork-io/terratest) library for infrastructure testing that do more in-depth integration tests of module functionality.

Reviewer: ```suggestion
- **[`terratest`](https://github.com/gruntwork-io/terratest) with chatops**: Our terraform tests are based on the `terratest` library for infrastructure testing that do more in-depth integration tests of module functionality.
```

These are run on request, and only by authorized contributors. We use ChatOps to trigger this workflow.

## Philosophy of Terraform Tests

At a minimum, we like to test to ensure that all of our modules cleanly `plan`, `apply` and `destroy`. This catches 80% of the problems with only 20% of the effort. We also want to test that the outputs are what we would expect. On the other hand, we don't want to retest what already is tested by HashiCorp in their providers. So for example, we have our [`terraform-aws-s3-bucket`](https://github.com/cloudposse/terraform-aws-s3-bucket) module that creates an S3 bucket. We don't need to test that a bucket is created; we assume that would be caught by the upstream terraform tests. But we do want to [test that the bucket name](https://github.com/cloudposse/terraform-aws-s3-bucket/blob/master/test/src/examples_complete_test.go#L38) is what we expect it to be, since this is an opinionated aspect of the module.

## Using ChatOps

> In order to securely run terraform tests and save on infrastructure costs we've setup some basic "ChatOps" style functionality. To use it you will need to have at least `triage` level of access to a repo. Typically, this is done by a CloudPosse team upon PR review.

Reviewer: ```suggestion
We've set up some basic "ChatOps" style functionality to securely run terraform tests and save on infrastructure costs. To use it, you must have at least the `triage` level of access to a repo. Typically, this is done by a CloudPosse team upon PR review.
```

Tests are initiated by posting GitHub comments on the PR. Currently supported commands are the following:

| Command                                             | Description                                                                        |
|-----------------------------------------------------|------------------------------------------------------------------------------------|
| `/terratest`                                        | Run the `terratest` integration tests in `test/src`                                |

Terraform tests run against our [testing infrastructure](https://github.com/cloudposse/testing.cloudposse.co) that we host in an isolated account on AWS, strictly for the purposes of testing infrastructure.

ChatOps is powered by [GitHub Actions](https://github.com/features/actions) and the [slash-dispatch-command](https://github.com/peter-evans/slash-command-dispatch).

The terratest workflow is defined in the [`cloudposse/actions`](https://github.com/cloudposse/actions/blob/master/.github/workflows/terratest-command.yml) repository. The benefit with this is that we have one place to control the testing

> workflow for all of our hundreds of terraform modules. The downside, however, with dispatched workflows is that the _workflows_ always run from the `master` branch.

Reviewer: We don't talk about the workflows, but should. Discuss the shared workflows and the general flow.

Reviewer: Sufficient that someone on the team could understand the various pieces and pick it up.

Author: It's a missing section here, I will add. 
The one above should stay as is until we get rid of dipatching terratest workflows to `cloudposse/action`

Author: @osterman Added shared workflow wording to the very top section

> ## ChatOps Configuration

Reviewer: ```suggestion
## ChatOps Initialization
```

> workflow for all of our hundreds of terraform modules. The downside, however, with dispatched workflows is that the _workflows_ always run from the `master` branch.

Reviewer 2: ```suggestion
workflow for all of our hundreds of terraform modules. The downside, however, with dispatched workflows is that the _workflows_ always run from the `main` branch.
```

## Terraform Testing

All of our Terraform modules have automated tests. We have two sets of checks: 

> - First set is running github shared workflow [`feature-branch`](https://github.com/cloudposse/github-actions-workflows-terraform-module/blob/main/.github/workflows/feature-branch.yml) 
> which does basic sanity checks: linting, formatting and docs. These checks are cheap to run and do not require any special permissions, so we run them automatically on every commit.

Reviewer 2: ```suggestion
- The first set of checks is executed through the feature-branch workflow, which can be found [here](https://github.com/cloudposse/github-actions-workflows-terraform-module/blob/main/.github/workflows/feature-branch.yml) 
This workflow generates some documentation and performs basic sanity checks, including linting and formatting. These checks are lightweight and can be executed without requiring any special permissions. Consequently, they are automatically run on every commit.
```

> You also can and should run this set of checks locally using `make precommit/terraform` command on your host before committing your changes locally and pushing your PR. You should run them because, among other things, they update the automatically generated documentation, and if you do not do it on your host, your PR will be blocked until you accept the changes via PR suggestions.

Reviewer 2: ~~~suggestion
Before committing and pushing your changes, you can and should run this set of checks locally by executing the following command on your host
```
make precommit/terraform
```
Running these checks locally incorporates all the required changes that otherwise would block your PR.
~~~

> - Second set runs terraform integration tests based on the [`terratest`](https://github.com/gruntwork-io/terratest) library for infrastructure testing that do more in-depth integration tests of module functionality.
> These are run on request, and only by authorized contributors. We use ChatOps to trigger this workflow.

Reviewer 2: ```suggestion
- The second set of checks consists of Terraform integration tests that validate the functionality and integration of the module. These tests are performed using the [`terratest`](https://github.com/gruntwork-io/terratest) library, specifically designed for infrastructure testing, and do more in-depth integration tests of module functionality.
Unlike the first set of checks, these integration tests are executed only on request, and only by authorized contributors. We use ChatOps to trigger this workflow.
```

## Philosophy of Terraform Integration Testing

> At a minimum, we like to test to ensure that all of our modules cleanly `plan`, `apply` and `destroy`. This catches 80% of the problems with only 20% of the effort. We also want to test that the outputs are what we would expect. On the other hand, we don't want to retest what already is tested by HashiCorp in their providers. So for example, we have our [`terraform-aws-s3-bucket`](https://github.com/cloudposse/terraform-aws-s3-bucket) module that creates an S3 bucket. We don't need to test that a bucket is created; we assume that would be caught by the upstream terraform tests. But we do want to [test that the bucket name](https://github.com/cloudposse/terraform-aws-s3-bucket/blob/master/test/src/examples_complete_test.go#L38) is what we expect it to be, since this is an opinionated aspect of the module.

Reviewer 2: ```suggestion
At a minimum, we ensure that all of our modules cleanly `plan`, `apply` and `destroy`. This catches 80% of the problems with only 20% of the effort. We also test than then the `enabled` input is set to `false`, no resources are created.

Ideally we would like to test that the resources are properly created, but often this is difficult to verify programmatically, in which case we settle for spot checking that the dynamic outputs match our expectations. At the same time, we do not want to waste effort retesting what has already been tested by HashiCorp and their providers. For example, we have our [`terraform-aws-s3-bucket`](https://github.com/cloudposse/terraform-aws-s3-bucket) module that creates an S3 bucket. We don't need to test that a bucket is created; we assume that would be caught by the upstream terraform tests. But we do want to [test that the bucket name](https://github.com/cloudposse/terraform-aws-s3-bucket/blob/master/test/src/examples_complete_test.go#L38) is what we expect it to be, since this is something under our control.
```

## Using ChatOps To Trigger Integration Tests

> In order to securely run terraform tests and save on infrastructure costs we've setup some basic "ChatOps" style functionality. To use it you will need to have at least `triage` level of access to a repo. Typically, this is done by a CloudPosse team upon PR review.

Reviewer 2: ```suggestion
In addition to automatic triggers, tests can be run on demand via "ChatOps". (You will need to have at least `triage` level of access to a repo to use ChatOps commands.) Typically, tests are run by a CloudPosse contributor or team member as part of a PR review.
```

- First set is running github shared workflow [`feature-branch`](https://github.com/cloudposse/github-actions-workflows-terraform-module/blob/main/.github/workflows/feature-branch.yml) 
which does basic sanity checks: linting, formatting and docs. These checks are cheap to run and do not require any special permissions, so we run them automatically on every commit.

> You also can run this set of checks locally using ```make precommit/terraform``` command on your host.

Reviewer 2: ```suggestion
You also can and should run this set of checks locally using `make precommit/terraform` command on your host before committing your changes locally and pushing your PR. You should run them because, among other things, it updates the automatically generated documentation, and if you do not do it on your host, your PR will be blocked until you accept the changes via PR suggestions.
```
